src/main/java/com/mercuryCyclists/Sales/service/OnlineSaleService.java

```diff
@@ -0,0 +1,92 @@
+import com.mercuryCyclists.Sales.repository.OnlineSaleRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+
+@Service
+public class OnlineSaleService {
+
+    private final OnlineSaleRepository onlineSaleRepository;
+    private final SaleService saleService;
+
+    @Autowired
+    public OnlineSaleService(OnlineSaleRepository onlineSaleRepository, SaleService saleService) {
+        this.onlineSaleRepository = onlineSaleRepository;
+        this.saleService = saleService;
+    }
+
+    /**
+     * Register a new Online Sale
+     */
+    public ResponseEntity<OnlineSale> registerOnlineSale(OnlineSale onlineSale) {
+        // Validate sale
+        if (!onlineSale.validate()) {
+            throw new IllegalStateException("Invalid Online Sale");
+        }
+
+        // Do after first validation in case quantity is null
+        if (onlineSale.getQuantity() <= 0) {
+            throw new IllegalStateException("Online Sale has 0 or a negative quantity");
+        }
+
+        // Check product exists
+        JsonObject product = saleService.getSaleProduct(onlineSale);
+        if(product.get("id") == null){
+            throw new IllegalArgumentException(String.format("Invalid product, %s", product));
+        }
+
+        // If there is enough of the product is stock
+        Long productQuantity = product.get("quantity").getAsLong();
+        Long saleQuantity = onlineSale.getQuantity();
+        if (productQuantity >= saleQuantity) {
+            // Update and save product
+            product.addProperty("quantity", (productQuantity - saleQuantity));
+            saleService.updateProduct(product);
+
+            // Save and return sale
+            onlineSaleRepository.save(onlineSale);
+            return new ResponseEntity<>(onlineSale, HttpStatus.CREATED);
+        } else {
+            // Get product parts
+            JsonArray productParts = saleService.getSaleProductParts(onlineSale);
+            ArrayList<JsonObject> partJsonObjs = new ArrayList<JsonObject>();
+
+            // For each part in the product
+            for (JsonElement part : productParts) {
+                // Convert json element to json object
+                JsonObject partJsonObj = part.getAsJsonObject();
+
+                Long partQuantity = partJsonObj.get("quantity").getAsLong();
+                if (partQuantity >= saleQuantity) {
+                    // Update part quantity
+                    partJsonObj.addProperty("quantity", (partQuantity - saleQuantity));
+
+                    // Save to ArrayList of JsonObjects
+                    partJsonObjs.add(partJsonObj);
+                } else {
+                    // return 303 Error
+                    return new ResponseEntity<>(onlineSale, HttpStatus.SEE_OTHER);
```

Comment:
Probably want to return a message as well redirecting to the backorder endpoint. IDK how to do this though, zhifa has done it for somthign

Reply:
done